### ISSUE TITLE
Issues with Packaging 

### ISSUE THREAD

**Issue by ddobric** [aspnet/dnx#2642]
Hi all,

just trying to publish ConsoleApp for ARM platform to execute it at Pi2. Unfortunately when application is started on device, it fails with following error:

PS C:\netcore\test> .\ConsoleApp
.\ConsoleApp : System.IO.FileLoadException: Could not load file or assembly 'System.Runtime, Version=4.0.21.0,
Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest
definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
    + CategoryInfo          : NotSpecified: (System.IO.FileL...LT: 0x80131040):String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

This console application references a some library called 'ConsoleLib', which provides some trivial functionality. When I remove the reference to 'ConsoleLib' application runs on device as expected.
It seems that packaging of referenced library didn't work correctly.

Interestingly when I publish the same application (with reference to ConsoleLib' for x86 and execute it, it works properly. But when I publish the same application for x64 dnx.exe crashes.

Assembly shown in error above 'system.runtime' is in packages folder as 4.0.21-beta-23220.

Here is what I use for build/publish:

Project.Json:
...
    "frameworks": {
        "dnx451": { },
        "dnxcore50": {  
       "dependencies": {
        "System.Collections": "4.0.10-beta-23019",
        "System.Console": "4.0.0-beta-23019",
        "System.Linq": "4.0.0-beta-23019",
        "System.Threading": "4.0.10-beta-23019",
        "Microsoft.CSharp": "4.0.0-beta-23019"
      } }  
    }

-Publish for ARM (crashes on PI2 with error shown above):
dnu publish --out "C:\abc" --no-source --runtime dnx-coreclr-win-arm.1.0.0-beta7-12364

-Publish for X64 (DNX.exe crashes):
dnu publish --out "C:\abc" --no-source --runtime dnx-coreclr-win-x64.1.0.0-beta5

-Publish for X86: Working version
dnu publish --out "C:\abc" --no-source --runtime dnx-coreclr-win-x86.1.0.0-beta8-15120

Thanks
Damir

**Comment by muratg** [comment]
Where are you getting dnx-coreclr-win-arm package from? I assume myget... 

cc @moozzyk

**Comment by moozzyk** [comment]
@ddobric - check in the project->properties->build of the ConsoleLib that you are building for AnyCPU and that the "Prefer 32-bit" flag is unchecked.

**Comment by ddobric** [comment]
Here is how I get runtimes: dnvm install latest -r coreclr -arch ARM -u

**Comment by ddobric** [comment]
When I take a look in global.json, which is produced in packaging process of X64, the architecture 'x86'. 

{
  "sdk": {
    "version": "1.0.0-beta8-15120",
    "runtime": "coreclr",
    "architecture": "x86"
  },
  "projects": [
    "src"
  ],
  "packages": "packages"
}

VS build props indicates .NETCore arch: x86. I cannot change it on anything else. Use DNX SDK beta8-*. But, for build I do not use VS at all. Where is this value set internally? I would expect it in project.Json.

This dependency story is a night mare, guys :)
Full packaging process should never produce output with missing references.

In general, I started with very trivial application which has a single line of code like: int a = 123;
To be able to compile this I added a reference to System.Runtime with some specific version.
All works fine, but on Pi2 doesn't work, because ARM runtime cannot resolve that runtime.
The I added reference with System.Runtime: "". This one does not build at all? Why?
Then I used System.Console as reference instead of System.Runtime. That worked well. It obviously resolved implicitly correct reference to System.Runtime.
But there is a still question: What is the way to setup and resolve references manually and correctly?
Resolving process must be much more transparent as we have it in .NET SDK. There are still many, many cases ,when assemblies are not resolved and there is no way to find out why.
This issue should not be integrated in .netcore :)

**Comment by muratg** [comment]
Can you do a `dnvm list`?  I'd like to see what version of DNX you're using.

**Comment by ddobric** [comment]
1.0.0-beta5       clr     x64          win
       1.0.0-beta5       clr     x86          win
       1.0.0-beta5       coreclr x64          win
       1.0.0-beta5       coreclr x86          win
       1.0.0-beta7-12364 coreclr arm          win
       1.0.0-beta8-15120 clr     x64          win
  *    1.0.0-beta8-15120 clr     x86          win             default
       1.0.0-beta8-15120 coreclr x86          win

**Comment by moozzyk** [comment]
@ddobric - you said that the problem occurrs if you reference a library called ConsoleLib. How do you build this library? How do you reference it? Is it an ASP.NET 5 library or just regular dll?

You can run only coreclr on arm. I am not sure if tooling supports arm at this point - last time I checked it did not. You can publish your application however from command line using `dnu publish`

**Comment by ddobric** [comment]
Pawel, ConsoleLib is just corenetlib created by VS template "class library (package)". 
My issue is related to correct references in project.json. I add there reference (dependency) to some assembly and  then package the application. When packaging for ARM and executing it on PI2, the application fails with "FileLoadException: Could not load file or assembly 'System.Runtime"..
This is probably nothing about PI2. It is more about adding dependencies and packaging.

1. How this error can happen after packaging process? Packing prepares all the app need.
2.Why DNX crashes when starting on x64 (after packaging)? It is probably bug caused by build for x86 and executing on x64. But again, packaging should make things correct way or throw an error.
3.What is the correct (best practices) way to peek right assemblies? (Using project.Json)

**Comment by moozzyk** [comment]
@ddobric - I think the issue is that you are trying to use recent CoreClr runtime packages (i.e.:
```
"System.Collections": "4.0.10-beta-23019",
"System.Console": "4.0.0-beta-23019",
"System.Linq": "4.0.0-beta-23019",
"System.Threading": "4.0.10-beta-23019",
"Microsoft.CSharp": "4.0.0-beta-23019"
```
with older runtimes - like beta5 or beta6. This won't work because you can't mix and match CoreClr packages and dnx runtimes. Unfortunately there is not a good way of telling what CoreClr versions should be used with what version of dnx runtimes - especially when you are using multiple feeds. For the latest unstable bits you could check versions here: https://www.myget.org/gallery/aspnetvnext (but in that case you could just probably just select the latest available version - as displayed by VS intellisense). Similary, if you just used the latest release version (i.e. beta7) you should use the latest version from the stable feed. You could also see the versions shipped as part of the latest stable release here https://www.myget.org/gallery/aspnetrelease

Publish just packages your application but it does not check if the application is actually able to run. If your application doesn't/can't run locally the packaged version will behave the same way (only, in some cases - e.g. arm published on non-arm, you can't check if it runs). I was able to repro the manifest issue (was wrong about Prefer 32-bit flag) by creating a beta7 HelloWorld application and then running it on beta5 coreclr runtime. I think fixing this bug #2197 may help a little bit in cases where a newer dnx runtime is being used with older CoreClr packages. 

The settings from the "sdk" section in the global.json file are currently used only by tooling.

**Comment by muratg** [comment]
@ddobric Did you try using matching runtimes with the packages that you use?  Is this resolved for you?

**Comment by ddobric** [comment]
All worked fine, after I used following references:
"System.Collections": "4.0.10-beta-23019",
"System.Console": "4.0.0-beta-23019",
"System.Linq": "4.0.0-beta-23019",
"System.Threading": "4.0.10-beta-23019",
"Microsoft.CSharp": "4.0.0-beta-23019"

I obviously have used some libraries, which are right now not supported at ARM: dnx-coreclr-win-arm.1.0.0-beta7-12364

But after all, it still remain few questions/suggestions:

1. How to find in which assembly (package) is which feature?
This is important when we know what feature we need, but do not know the package. For example: Where is implemented Console.WriteLine()?

2. We need a more transparent way to enlist/document features implemented in the assembly(package).
This is important when we don't know features and looking for them in various assemblies (packages).
For example: I'm looking for output functionality in all packages and don't know that is implemented in Console.WriteLine().

3. How find out which runtime does provide required functionalities?
This is the case which lead me to start this post. I solved 1 and 2, but failed after deployment, by trying to start application.
For example: dnu publish MUST check that all references CAN BE resolved. 

Thanks for your help and hope these few suggestions summarize tooling and organizational gaps.
:)
